src/ansi-c/c_typecheck_base.cpp

```diff
@@ -645,6 +645,10 @@ void c_typecheck_baset::typecheck_declaration(
     irept contract;
 
     {
+      exprt spec_assigns =
+        static_cast<const exprt &>(declaration.find(ID_C_spec_assigns));
```

Comment:
Are we guaranteed this cast is safe?

Reply:
While find returns an irept object, which is a parent of exprt, the way the parser is currently implemented, we should always be getting an exprt when searching for ID_C_spec_assigns. If someone does something strange in the parser, we might encounter a problem, but I think this is very unlikely.

Reply:
I added `spec_assigns` to `ansi_c_declarationt` for assigns clauses, ensures and requires. We should add a single exprt subclass with accessors, but this is not specific to assigns clauses. We'd need to replace all casts for ensures and requires as well. Thus, I recommend doing so on a separate PR specific for this clean up.